pkg/typesystem/resolver.go

```diff
@@ -0,0 +1,79 @@
+package typesystem
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/karlseguin/ccache/v3"
+	"github.com/oklog/ulid/v2"
+	"github.com/openfga/openfga/pkg/storage"
+)
+
+const (
+	typesystemCacheTTL = 168 * time.Hour // 7 days
+)
+
+// TypesystemResolverFunc is a function that implementations can implement to provide lookup and
+// resolution of a Typesystem.
+type TypesystemResolverFunc func(ctx context.Context, storeID, modelID string) (*TypeSystem, error)
+
+// MemoizedTypesystemResolverFunc returns a TypesystemResolverFunc that either fetches the provided authorization
+// model (if provided) or looks up the latest authorizaiton model, and then it constructs a TypeSystem from
+// the resolved model. The type-system resolution is memoized so if another lookup of the same model occurs,
+// then the earlier TypeSystem that was constructed will be used.
+//
+// The memoized resolver function is safe for concurrent use.
+func MemoizedTypesystemResolverFunc(datastore storage.AuthorizationModelReadBackend) TypesystemResolverFunc {
+
+	cache := ccache.New(ccache.Configure[*TypeSystem]())
```

Comment:
I won't block this PR, but i don't feel super happy about having this cache here and one for `AuthorizationModels` [here](https://github.com/openfga/openfga/blob/5c7c5c88dd07c85666275621503f867428a59693/pkg/storage/caching/caching.go#L21).

Reply:
I'd be happy to remove the code in https://github.com/openfga/openfga/blob/main/pkg/storage/caching/caching.go and move it over into here.

@elbuo8 already suggested we also integrate `singleflight into this work, and so I can just do that now and then delete that whole `storage/caching` package.

How does that sound to you?

Reply:
Yeah that'd be nice :D

Reply:
Ok, done 👍 

Tell me what you think.

Reply:
I think we introduce a regression by removing `caching.go`. Because for example, `server.Expand` calls `s.resolveTypesystem()` which fetches the auth model, and then `NewExpandQuery.Execute()` calls `datastore.ReadAuthorizationModel` which will fetch the model again, no? 

If we remove `caching.go` we'd need to pass `typesys` in all `Execute()` calls, which is a reasonable path forward but it will blow up this PR with changes

Reply:
Yeah, I guess that's why I need to revive and finish https://github.com/openfga/openfga/pull/756 after this PR is merged.

I'll revert the deletion of `pkg/storage/caching` and then we can drop it later.

Reply:
@miparnisari ok I've reverted the commit that deletes the `pkg/storage/caching` code, but I've integrated the `singleflight` work into the memorized typesystem resolver. That should do it.